### ISSUE TITLE
[BUG] Exits without a destination are treated as Objects

### ISSUE THREAD

**Issue by strikaco** [evennia/evennia#2362]
#### Describe the bug
Exits that do not have a destination are treated as Objects in room descriptions and elsewhere.

```
Limbo(#2)
Welcome to your new Evennia-based game! Visit http://www.evennia.com if you need
help, want to contribute, report issues or just join the community.
As Account #1 you can create a demo/tutorial area with 'batchcommand tutorial_world.build'.
Exits: down(#7)
You see: an exit(#3)
```

(`down` was created using `tunnel`. `exit` was created from a prototype that did not specify a destination.)

`#3` _is_ a child of `Exit`:

```
Name/key: exit (#3)
Aliases: an exit(#3),one exit(#3)
Typeclass: Exit (typeclasses.exits.Exit)
```

This also impacts the `exits` property of Room objects:

```
>>> self.location.exits
[down]
```

**Comment by fariparedes** [comment]
Is this really a bug? If an exit has no destination, you can't exactly use it as an exit, can you?

**Comment by strikaco** [comment]
> Is this really a bug? If an exit has no destination, you can't exactly use it as an exit, can you?

By this logic any exit you can't traverse should also receive the same treatment. A locked exit can't be used, but we still consider it an exit.

In my case I'm literally trying to make a "fake" door. It looks silly being counted amongst random garbage in the room.

**Comment by kovitikus** [comment]
An object is categorized as an exit by simply having a destination. This is hinted at many times throughout the documentation of Evennia's source code. I'm not going to claim it's the only characteristic that defines an exit, but it is certainly a critical factor.

***
Here's one such example. I've seen it elsewhere, in some sort of object creation method, but I can't recall where.

https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L293
```
Returns all exits from this object, i.e. all objects at this
        location having the property destination != `None`.
```

**Comment by strikaco** [comment]
That's a dirty shortcut used in favor of the more-expensive `inherits_from` call. Every object with `location=None` used to be considered a Room in certain contexts, too. It's misleading.

Giving some object a destination does not make it a functional exit either. It's actually _more_ of an abomination because although an object-with-a-destination is presented as an exit and can be looked at, it raises "command not found" if you try to _use_ it as the exit it claims to be. 

I get what you two are saying (you're taking the ECS approach), but typeclasses are supposed to be object-oriented-- if something's typeclass is explicitly defined as `typeclasses.Exit`, second-guessing that based on arbitrary behaviors makes the typeclass declaration itself meaningless.

**Comment by kovitikus** [comment]
Typeclasses are overrated anyways. I'm pro exit-fluidity.

I've been using tags to define behaviors of my objects, such as the type of exit (door, stairs, etc). It's probably a bad idea and I can't imagine the mess it'll make when two conflicting tags get added to a single object, but it's easier than checking for and setting typeclass paths in my code (which breaks if I change anything regarding that path anyways).

*I recently refactored the structure of my project and it was a nightmare where I had hardcoded any typeclass paths and checks.*

In any case, I agree that there's far more to an exit than just the destination. For me though, I set my own [custom movement commands](https://github.com/kovitikus/hecate/blob/master/travel/travel_cmds.py#L96), so the ones generated by the exit aren't that big of a deal.

***

As far as the original post of this thread, are you suggesting that an exit object with no destination be listed as a valid exit or simply not listed as an object and omitted from the look command altogether? The one with no destination is in a sort of limbo state. I assume it is still of the Exit typeclass, but look filters the `return_appearance` of the room by checking an object for a destination.

https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L1657

***

It's been [discussed](https://irclogs.jackgrigg.com/irc.freenode.net/evennia/2019-07-16#i_4446534) before that the return_appearance hook is far too bulky and could use a refactoring and this seems especially true in regards to your complaint about the limbo exit object you spawned.

**Comment by TehomCD** [comment]
What makes the most sense to me would be adding a method that allows a given typeclass to specify a string type for categorization/announcement purposes, which would then be used by the return_appearance method or elsewhere. Why not use duck typing here when we're already using it with typeclasses elsewhere? Call it `get_typeclass_category_name`, return "Exit" as a default for exits, "Room" as a default for room, etc.

**Comment by kovitikus** [comment]
> 
> 
> What makes the most sense to me would be adding a method that allows a given typeclass to specify a string type for categorization/announcement purposes, which would then be used by the return_appearance method or elsewhere. Why not use duck typing here when we're already using it with typeclasses elsewhere? Call it `get_typeclass_category_name`, return "Exit" as a default for exits, "Room" as a default for room, etc.

Yeah, that would help a lot with the return_appearance hook, which isn't super intelligent. 

*Coincidentally enough, it was one of the easiest bits of code for me to read and edit when I first started as well.*

**Comment by strikaco** [comment]
> Typeclasses are overrated anyways. I'm pro exit-fluidity.
> 
> I've been using tags to define behaviors of my objects, such as the type of exit ([door, stairs, etc](https://github.com/kovitikus/hecate/blob/master/travel/travel_handler.py#L166)). It's probably a bad idea and I can't imagine the mess it'll make when two conflicting tags get added to a single object, but it's easier than checking for and setting typeclass paths in my code (which breaks if I change anything regarding that path anyways).

I get the impression the objects/rooms/exits/characters(mobiles) typeclass taxonomy is a Diku holdover. These four object types are very distinct in terms of their abstract functionality and are a pretty sensible way of classifying generic objects and spaces. You don't really puppet exits, carry rooms, enter objects or get characters, and namespacing this way prevents that sort of illogical behavior.

Using tags to define behaviors also happens to be a Diku trait, so you're not in uncharted territory. I'm going this direction myself; it helps to namespace tags per-object-type using the category field. The "underwater[room]" tag is distinct from "underwater[mobile]." One identifies a room that is underwater, slows movement and causes periodic drowning; the other identifies a mobile which can operate normally in such an environment. (The last time I tried this, I had a lot of problems with _something_ clobbering the `db_data` field on the tag objects and deleting my metadata.)

> As far as the original post of this thread, are you suggesting that an exit object with no destination be listed as a valid exit or simply not listed as an object and omitted from the look command altogether? The one with no destination is in a sort of limbo state. I assume it is still of the Exit typeclass, but look filters the `return_appearance` of the room by checking an object for a destination.
> 
> https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L1657

If it's a descendant of `typeclasses.Exit`, it should be listed with the exits-- whether it has a destination or not. Having a destination should not be the single factor that determines whether something is an exit, since (as above) objects with a destination end up being _less_ functional than my door with no destination does.

The use cases for this would be things like fake doors or balconies, where you want it to "look" like the character can go in a certain direction per the room description, but if they look or try to go that way they get a description of an obstacle instead. The end result is that they don't leave the room.

> It's been [discussed](https://irclogs.jackgrigg.com/irc.freenode.net/evennia/2019-07-16#i_4446534) before that the return_appearance hook is a bit wonky and could use a refactoring. This seems especially true in regards to your complaint about the limbo exit object you spawned.

Return_appearance certainly needs an overhaul, but I don't think that is a factor in this case-- see the `exits` property on Room objects.

**Comment by fariparedes** [comment]
> 
> 
> > Is this really a bug? If an exit has no destination, you can't exactly use it as an exit, can you?
> 
> By this logic any exit you can't traverse should also receive the same treatment. A locked exit can't be used, but we still consider it an exit.
> 
> In my case I'm literally trying to make a "fake" door. It looks silly being counted amongst random garbage in the room.

Not at all; a superuser can still go through a locked exit since they bypass all locks. But an exit with no destination is totally useless as an exit. But since you brought it up, maybe just making an exit with a `traverse:false()` lock would solve your needs?

**Comment by strikaco** [comment]
> > > Is this really a bug? If an exit has no destination, you can't exactly use it as an exit, can you?
> > 
> > 
> > By this logic any exit you can't traverse should also receive the same treatment. A locked exit can't be used, but we still consider it an exit.
> > In my case I'm literally trying to make a "fake" door. It looks silly being counted amongst random garbage in the room.
> 
> Not at all; a superuser can still go through a locked exit since they bypass all locks. But an exit with no destination is totally useless as an exit. But since you brought it up, maybe just making an exit with a `traverse:false()` lock would solve your needs?

Thanks, I ended up setting the destination to self.location, then making it non-traversible as you suggest.

**Comment by Griatch** [comment]
I'll take a look at this later, it needs some further thinking about.
